src/sparseml/optim/helpers.py

```diff
@@ -569,6 +574,44 @@ def _check_warn_dict_difference(original_dict, new_dict):
     return new_dict
 
 
+def add_framework_metadata(metadata: Dict[str, Dict]) -> Dict[str, Dict]:
+    """
+    Adds the information about the relevant frameworks used by the user to the metadata.
+    :param metadata: Validated metadata
+    :return: Validated metadata with framework metadata
+    """
+
+    framework_metadata = {
+        "python_version": platform.python_version(),
+        "torch_version": torch.__version__,
+        "sparseml_version": sparseml_version,
+    }
+    for stage_name, stage_value in metadata.items():
+        if stage_value is None:
+            _stage_value = framework_metadata
+        else:
+            shared_keys = set(stage_value.keys()).intersection(
+                set(framework_metadata.keys())
+            )
+            if shared_keys:
+                warning_if_stage = (
+                    f"stage, stage name: {stage_name}"
+                    if stage_name != RECIPE_METADATA_KEY
+                    else ""
+                )
+                warning_msg = (
+                    f"Overwriting metadata {warning_if_stage} key(s) "
+                    f"{shared_keys} with new value(s) "
+                    f"{ {k:v for k,v in framework_metadata.items() if k in shared_keys} }"  # noqa E501
+                )
+                warnings.warn(warning_msg)
```

Comment:
nit: use logger.warning instead of warnings

Reply:
good catch, i will apply it to `_check_warn_dict_difference` as well.